### DISCUSSION TITLE
subifd and directories using libtiff

### DISCUSSION THREAD

**Discussion by rsmith22** [libvips/libvips#3128] [Q&A]
Hello

I am using libtiff to write pyramid images, however, I am required to add subifds for the pyramid resolutions to satisfy Bioformats style pyramid images.

The libtiff code examples for this are not very useful. My code will generate pyramid images but the directories are not linked to the subifd  array set with the main image using TiffTag.SUBIFD. TheTiffTag.Subfiletype Reduced is set for the pyramid image where the subifd's should link to their directory.

Could you help me resolve this matter as I understand libvips has this capability.

**Comment by jcupitt** [comment]
Hi @rsmith22,

Yes, that took me a while to work out, the tiff docs are not very clear. 

The libvips code is here:

https://github.com/libvips/libvips/blob/master/libvips/foreign/vips2tiff.c#L2113-L2142

You can try out the libvips pyramid generator like this:

```
$ vips copy astronauts.png astronauts.tiff[pyramid,tile,subifd]
$ tiffinfo astronauts.tiff 
TIFF Directory at offset 0x3145736 (300008)
  Image Width: 900 Image Length: 900
  Tile Width: 128 Tile Length: 128
  Resolution: 10, 10 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
  SubIFD Offsets: 3932802 4129726 4179122
$ tiffinfo astronauts.tiff -o 3932802
TIFF Directory at offset 0x3932802 (3c0282)
  Subfile Type: reduced-resolution image (1 = 0x1)
  Image Width: 450 Image Length: 450
  Tile Width: 128 Tile Length: 128
  Resolution: 10, 10 pixels/cm
  Bits/Sample: 8
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
```

So the SUBIFD offsets are linked to the pyramid layers.

**Reply by rsmith22** [reply]
Hello 

I found this in your code and implemented it however my code's pyramid tiffs do not contain subifd offsets. I believe I followed your code and it is similar to the thumbnail example below, in which a write directory occurs after the main image and write directory occurs after each reduced image. I thought the libtiff took care of the subifd offset calculations with this approach. However, something isn't correct. Am I interpreting your code correctly?
https://www.appsloveworld.com/cplus/100/85/in-a-tiff-create-a-sub-ifd-with-thumbnail-libtiff

The information from my generated pyramid tiff is as follows

OMETiffReader initializing C:\CellDiveImages\SLIDE_30_1.0.1_R000_DAPI_AF_F.omeT3X.ome.tif
Reading IFDs
Populating metadata
Reading IFDs
Populating metadata
Initialization took 0.11s

Reading core metadata
filename = SLIDE_30_1.0.1_R000_DAPI_AF_F.omeT3X.ome.tif
Used files = [SLIDE_30_1.0.1_R000_DAPI_AF_F.ome.tif]
Series count = 8
Reading IFDs
Populating metadata
Series #0 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0

Reading IFDs
Populating metadata
Series #1 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0

Reading IFDs
Populating metadata
Series #2 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0

Reading IFDs
Populating metadata
Series #3 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0

Reading IFDs
Populating metadata
Series #4 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0

Reading IFDs
Populating metadata
Series #5 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0

Reading IFDs
Populating metadata
Series #6 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0

Reading IFDs
Populating metadata
Series #7 :
        Image count = 1
        RGB = false (1)
        Interleaved = false
        Indexed = false (false color)
        Width = 11285
        Height = 5768
        SizeZ = 1
        SizeT = 1
        SizeC = 1
        Tile size = 512 x 512
        Thumbnail size = 128 x 65
        Endianness = intel (little)
        Dimension order = XYCZT (certain)
        Pixel type = uint16
        Valid bits per pixel = 16
        Metadata complete = true
        Thumbnail series = false
        -----
        Plane #0 <=> Z 0, C 0, T 0


Reading global metadata
BitsPerSample: 16
Compression: LZW
ImageLength: 5768
ImageWidth: 11285
MetaDataPhotometricInterpretation: Monochrome
MetaMorph: no
NumberOfChannels: 1
Orientation: 1st row - top; 1st column - left
PhotometricInterpretation: BlackIsZero
PlanarConfiguration: Chunky
Predictor: Horizontal differencing
ResolutionUnit: Centimeter
SamplesPerPixel: 1
TileByteCounts: 528591
TileLength: 512
TileWidth: 512
XResolution: 30769.327117332683
YResolution: 30769.327117332683

**Reply by jcupitt** [reply]
Welcome to fighting image format libraries :( Sorry, you'll just have to keep tinkering until it works.

I would make a minimal working example derived from the known-good libvips writer, then build that up into a solution for your code.

**Reply by rsmith22** [reply]
Will do, can you let me know what version of libtiff you are using and whether you made changes to it?

Thank you

**Reply by jcupitt** [reply]
Just regular libtiff 4.x.

**Reply by rsmith22** [reply]
with no changes? The reason I ask, an example on the internet explains method to add to libtiff which calculates and sets the subifd offsets.

**Reply by jcupitt** [reply]
That example is not correct -- you just need to set the ifds to 0 and libtiff will calculate them for you.

**Reply by rsmith22** [reply]
thank you for checking
